strawberryfields/backends/tfbackend/circuit.py

```diff
@@ -709,140 +669,132 @@ def measure_homodyne(self, phi, mode, select=None, **kwargs):
+                        for n, m in number_state_indices]
+            hermite_matrix = tf.scatter_nd(number_state_indices, terms, [self._cutoff_dim, self._cutoff_dim, num_bins])
+            hermite_terms = tf.multiply(tf.expand_dims(reduced_state, -1), tf.expand_dims(tf.cast(hermite_matrix, ops.def_type), 0))
+            rho_dist = tf.cast(tf.reduce_sum(hermite_terms, axis=[1, 2]), tf.float64) \
+                                    * (m_omega_over_hbar / np.pi) ** 0.5 \
+                                    * tf.exp(- x ** 2) \
+                                    * (q_tensor[1] - q_tensor[0]) # Delta_q for normalization (only works if the bins are equally spaced)
+
+            # use tf.random.categorical to sample
+            logprobs = tf.math.log(rho_dist)
+            samples_idx = tf.random.categorical(logprobs, 1)
+            homodyne_sample = tf.gather(q_tensor, samples_idx)
+            homodyne_sample = tf.squeeze(homodyne_sample)
+
+        meas_result = tf.identity(homodyne_sample, name="Meas_result")
+
+        # project remaining modes into conditional state
+        if self._num_modes == 1:
+            # in this case, all modes were measured and we we put everything into vacuum
+            self.reset(pure=self._state_is_pure)
+        else:
+            # only some modes were measured: put unmeasured modes in conditional state, while reseting measured modes to vac
+            inf_squeezed_vac = tf.convert_to_tensor([(-0.5) ** (m // 2) * np.sqrt(factorial(m)) / factorial(m // 2) if m % 2 == 0 else 0. for m in range(self._cutoff_dim)],
+                                                    dtype=ops.def_type)
+            if self._batched:
+                inf_squeezed_vac = tf.tile(tf.expand_dims(inf_squeezed_vac, 0), [batch_size, 1])
+            displacement_size = tf.stack(tf.convert_to_tensor(meas_result * np.sqrt(m_omega_over_hbar / 2)))
+            quad_eigenstate = ops.displacement(displacement_size, 0, inf_squeezed_vac, self._cutoff_dim, True, self._batched)
+            homodyne_eigenstate = ops.phase_shifter(phi, 0, quad_eigenstate, self._cutoff_dim, True, self._batched)
 
-                conditional_state = ops.conditional_state(self._state, homodyne_eigenstate, mode, self._state_is_pure, batched=self._batched)
+            conditional_state = ops.conditional_state(self._state, homodyne_eigenstate, mode, self._state_is_pure, batched=self._batched)
 
-                # normalize
-                if self._state_is_pure:
-                    norm = tf.norm(tf.reshape(conditional_state, [batch_size, -1]), axis=1)
+            # normalize
+            if self._state_is_pure:
+                norm = tf.norm(tf.reshape(conditional_state, [batch_size, -1]), axis=1)
+            else:
+                # calculate norm of conditional_state
+                # cheap hack since tensorflow doesn't allow einsum equation for trace:
```

Comment:
Is this supported in TF2 now?

Reply:
Apparently not yet, I tried on a small example and I get:
`ValueError: Subscript not supported: an axis appears more than once: ijkkll`